src/libs/blueprint/conduit_blueprint_mesh_utils.hpp

```diff
@@ -215,6 +215,205 @@ CONDUIT_BLUEPRINT_API const Node * find_reference_node(const Node &node, const s
+    ///     - offset (optional) specifies where the data starts in the array
+    ///     - stride (optional) specifies the extent of the array storing data
+    ///
+    /// The shape is required.  It is an error if shape is omitted.
+    /// If offset is not specified, it defaults to 0 in each dimension.
+    /// If stride is not specified, it defaults to
+    /// \code
+    /// stride[0] = 1
+    /// stride[i] = stride[i-1] * (offset[i-1] + shape[i-1])
+    /// \endcode
+    /// Node that offset is specified in terms of logical index, not
+    /// flatindex, and stride is specified in terms of flatindex.  Also note
+    /// that the default stride holds an assumption that the data is laid
+    /// out in C-style, with fastest-varying dimension left-most.  Users may
+    /// specify a custom stride to index Fortran-style arrays, where the
+    /// fastest-varying index is right-most.
+    ///
+    /// Here are a few examples:
+    ///
+    /// - A 6x4 array
+    ///   \code
+    ///   shape: [6, 4]
+    ///   \endcode
+    /// - A 6x4 array with two extra elements at the end of each row
+    ///   \code
+    ///   shape: [6, 4]
+    ///   stride: [1, 8]
+    ///   \endcode
+    /// - A 6x4 array with two elements of padding on the low end of a
+    ///   dimension and one element of padding on the high end
+    ///   \code
+    ///   shape: [6, 4]
+    ///   offset: [2, 2]
+    ///   stride: [1, 9]
+    ///   \endcode
+    /// - A 6x4x5 array with two elements of padding on the low end of each
+    ///   dimension and one element of padding on the high end (adds third
+    ///   dimension to previous)
+    ///   \code
+    ///   shape: [6, 4, 5]
+    ///   offset: [2, 2, 2]
+    ///   stride: [1, 9, 63]
+    ///   \endcode
+    /// - A Fortran 6x4x5 array with two elements of padding on the low
+    ///   end of each dimension and one element of padding on the high end
+    ///   (previous example changed to column-major)
+    ///   \code
+    ///   shape: [6, 4, 5]
+    ///   offset: [2, 2, 2]
+    ///   stride: [63, 7, 1]
+    ///   \endcode
+    NDIndex(const Node& node);
+
+    /// Array constructor
+    NDIndex(const index_t dim, const index_t* shape, const index_t* offset = NULL, const index_t* stride = NULL);
+
+    /// Destructor
+    ~NDIndex() { };
+
+    /// Assignment operator.
+    NDIndex& operator=(const NDIndex& itr);
+
+    //-----------------------------------------------------------------------------
+    /// Retrieve a flat-index: public interface.
+    //-----------------------------------------------------------------------------
+    template<typename T, typename... Ts>
+    index_t     index(T idx, Ts... idxs) const;
+    template<typename T>
+    index_t     index(T idx) const;
+
+    /// With default argument, returns the number of ranks or dimensions for
+    /// this NDIndex.  With dim >= 0, returns the extent of this NDIndex
+    /// for dimension dim.
+    index_t     shape(index_t dim = -1) const;
```

Comment:
instead of magic number here, can we just add a dims() method?

Reply:
Sure. We had discussed shape(-1) returning the extent.  Dims() is fine.  I think I like rank() better, though.

Reply:
Yes, in hindsight it just looks too magical. Calling equiv: `.shape()` seems better, but still magical. 

rank would match fortran vernacular, correct? 

Does RAJA have a term for dims / rank/ etc?

Are there other sources of inspiration? 

What does Axom use? 

Note sure if numpy helps b/c folks just take len(v.shape) -- I don't think there is explicit method to ask this?

Reply:
Actually: numpy has:

`v.ndims`